src/operator/mkldnn/mkldnn_pooling-inl.h

```diff
@@ -0,0 +1,220 @@
+/*!
+ * Copyright (c) 2015 by Contributors
+ * \file mkldnn_pooling-inl.h
+ * \brief
+ * \author yandai
+*/
+
+#ifndef MXNET_OPERATOR_MKLDNN_MKLDNN_POOLING_INL_H_
+#define MXNET_OPERATOR_MKLDNN_MKLDNN_POOLING_INL_H_
+#include <vector>
+#include <algorithm>
+#include "../pooling-inl.h"
+#include "mkldnn_memory-inl.h"
+
+namespace mxnet {
+namespace op {
+
+template <typename xpu, typename Reducer, typename DType>
+class MKLDNNPoolingOp : public PoolingOp<xpu, Reducer, DType> {
+ public:
+  explicit MKLDNNPoolingOp(PoolingParam p)
+      : PoolingOp<xpu, Reducer, DType>(p),
+        support_mkldnn_(false),
+        init_mkldnn_(false),
+        poolingFwd_(NULL),
+        poolingBwd_(NULL),
+        fwd_out_data_(new MKLData<DType>()),
+        fwd_in_data_(new MKLData<DType>()),
+        bwd_in_diff_(new MKLData<DType>()),
+        bwd_out_diff_(new MKLData<DType>()) {
+    switch (this->param_.pool_type) {
+      case pool_enum::kMaxPooling:
+        mode_ = dnnAlgorithmPoolingMax;
+        break;
+      case pool_enum::kAvgPooling:
+        mode_ = dnnAlgorithmPoolingAvg;
+        break;
+      default:
+        LOG(FATAL) << "Not implmented";
+    }
+  }
+
+  ~MKLDNNPoolingOp() {
+    dnnDelete<DType>(poolingFwd_);
+    dnnDelete<DType>(poolingBwd_);
+    if (support_mkldnn_) {
+      dnnReleaseBuffer<DType>(pooling_res_[dnnResourceWorkspace]);
+    }
+  }
+
+  virtual void Forward(const OpContext &ctx, const std::vector<TBlob> &in_data,
+                       const std::vector<OpReqType> &req,
+                       const std::vector<TBlob> &out_data,
+                       const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(in_data.size(), 1);
+    CHECK_EQ(out_data.size(), 1);
+    Stream<cpu> *s = ctx.get_stream<cpu>();
+    if (this->param_.kernel.ndim() == 2) {
+      // 2d pool
+      if (!init_mkldnn_) {
+        this->Init(s, in_data, out_data);
+      }
+      if (support_mkldnn_) {
+        Tensor<cpu, 4, DType> data =
+            in_data[pool_enum::kData].get<cpu, 4, DType>(s);
+        Tensor<cpu, 4, DType> out =
+            out_data[pool_enum::kOut].get<cpu, 4, DType>(s);
+        pooling_res_[dnnResourceSrc] =
+            reinterpret_cast<void *>(fwd_in_data_->get_converted_prv(data.dptr_, false));
+        pooling_res_[dnnResourceDst] =
+            reinterpret_cast<void *>(fwd_out_data_->get_converted_prv(out.dptr_, false));
+        CHECK_EQ(dnnExecute<DType>(poolingFwd_, pooling_res_), E_SUCCESS);
+        fwd_out_data_->get_output_ptr(data.dptr_);
+      } else {
+        PoolingOp<cpu, Reducer, DType>::Forward(ctx, in_data, req, out_data,
```

Comment:
Can you determine support_mkldnn_ at operator creation time? If so please directly create a PoolingOp instead of forwarding inside mkldnn_pooling